File: DanceLib/DanceValidation.cs
using Newtonsoft.Json;

namespace DanceLibrary;

/// <summary>
/// Validation rules for tempo and meter from external sources (Spotify/EchoNest).
/// Used to detect and correct common tempo detection errors.
/// </summary>
public class DanceValidation
{
    /// <summary>
    /// If tempo is below this threshold, double it (half-time detection error).
    /// </summary>
    [JsonProperty("doubleTempoIfBelow", NullValueHandling = NullValueHandling.Ignore)]
    public decimal? DoubleTempoIfBelow { get; set; }

    /// <summary>
    /// If tempo is above this threshold, halve it (double-time detection error).
    /// </summary>
    [JsonProperty("halveTempoIfAbove", NullValueHandling = NullValueHandling.Ignore)]
    public decimal? HalveTempoIfAbove { get; set; }

    /// <summary>
    /// List of meter signatures (e.g., "3/4", "6/8") that should trigger a flag for manual review.
    /// </summary>
    [JsonProperty("flagInvalidMeters", NullValueHandling = NullValueHandling.Ignore)]
    public List<string> FlagInvalidMeters { get; set; }
}

File: m4d/ClientApp/src/models/DanceDatabase/DanceValidation.ts
import { jsonArrayMember, jsonMember, jsonObject } from "typedjson";

// Mirrors DanceLib/DanceValidation.cs - sanity-check thresholds used to catch Spotify/EchoNest
// half-time/double-time tempo detection errors. Lives on DanceType (dances.json's top-level
// "validation" key), not per style/instance - see DanceType.Validation in DanceLib for why.
@jsonObject
export class DanceValidation {
  @jsonMember(Number) public doubleTempoIfBelow?: number;
  @jsonMember(Number) public halveTempoIfAbove?: number;
  @jsonArrayMember(String) public flagInvalidMeters?: string[];
}
